### PR TITLE
V4.0 replication (WIP)

### PR DESCRIPTION
**Pull request by myarichuk**

### CODE REVIEW

File: src/Raven.Server/Documents/SqlReplication/SqlReplicationLoader.cs [hunk 1]
```diff
@@ -1,71 +1,42 @@
...
-
-                if (Log.IsDebugEnabled)
-                    Log.Debug(() => $"Sql Replication configuration was changed: {notification.Key}");
-            }
         }
 
-        public void Initialize()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        protected override bool ShouldReloadConfiguration(string systemDocumentKey)
         {
```
Comment (ayende) [inline src/Raven.Server/Documents/SqlReplication/SqlReplicationLoader.cs:32]: AggressiveInlining and virtual method calls don't go together

File: src/Raven.Client/Connection/Async/AsyncAdminServerClient.cs [hunk 2]
```diff
@@ -49,9 +49,16 @@ public AsyncAdminServerClient(AsyncServerClient asyncServerClient)
...
             {
-                await req.ExecuteRequestAsync().WithCancellation(token).ConfigureAwait(false);
+                using (var req = adminRequest.DeleteDatabase(databaseName, hardDelete))
+                {
+                    await req.ExecuteRequestAsync().WithCancellation(token).ConfigureAwait(false);
+                }
+            }
+            catch (Exception e)
+            {
+                throw;
```
Comment (ayende) [inline src/Raven.Client/Connection/Async/AsyncAdminServerClient.cs:61]: why?

Comment (myarichuk) [inline src/Raven.Client/Connection/Async/AsyncAdminServerClient.cs:61]: Hmm. leftover of debugging I think. I will remove in next PR

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 3]
```diff
@@ -96,8 +103,10 @@ private void InitializeInternal()
         public void Dispose()
         {
             _databaseShutdown.Cancel();
+            DocumentReplicationLoader.Dispose();
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentDatabase.cs:106]: Why isn't this using the exception aggregator?

File: src/Raven.Server/Documents/SqlReplication/SqlReplicationLoader.cs (hunk 1 above)
Comment (myarichuk) [inline src/Raven.Server/Documents/SqlReplication/SqlReplicationLoader.cs:32]: removed for next PR, thanks

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 4]
```diff
@@ -657,15 +660,20 @@ public bool DeleteCollection(DocumentsOperationContext context, string name, Lis
 
             var newEtag = ++_lastEtag;
             var newEtagBigEndian = IPAddress.HostToNetworkOrder(newEtag);
-
+                    
+            var changeVector = _documentDatabase.
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: This is wrong

Comment (myarichuk) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: I've commented this in later commit, will delete that in next PR

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 3 above)
Comment (myarichuk) [inline src/Raven.Server/Documents/DocumentDatabase.cs:106]: no reason; good point, fixed it.

File: src/Raven.Server/Documents/SqlReplication/SqlReplication.cs [hunk 5]
```diff
@@ -272,7 +272,6 @@ public override void Dispose()
         {
             _cancellationTokenSource.Cancel();
             _disposed = true;
-            base.Dispose();
```
Comment (ayende) [inline src/Raven.Server/Documents/SqlReplication/SqlReplication.cs]: Why aren't you calling the base class dispose?

**Pull request merged**